### PR TITLE
Implementation of pay wrapper

### PR DESCRIPTION
**Pull request by talbaneth**

### CODE REVIEW

File: contracts/KyberPayWrapper.sol [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        address walletId,
+        bytes hint)
+    external
+    payable
+    returns(uint);
+}
+
+
+contract KyberPayWrapper is Withdrawable, ReentrancyGuard {
+    ERC20 constant public ETH_TOKEN_ADDRESS = ERC20(0x00eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee);
```
Comment (yaronvel) [inline contracts/KyberPayWrapper.sol:26]: I think this constant is already defined in `Withdrawable`

File: contracts/KyberPayWrapper.sol [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    {
+
+        require(src != address(0));
+        require(dest != address(0));
+        require(destAddress != address(0));
+
+        if (src == ETH_TOKEN_ADDRESS) require(srcAmount == msg.value);
+
+        PayData memory payData = PayData(
+            src,
```
Comment (yaronvel) [inline contracts/KyberPayWrapper.sol]: use amit notation

File: contracts/KyberPayWrapper.sol [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            walletId,
+            paymentData,
+            hint,
+            kyberNetworkProxy
+        );
+
+        uint paidAmount = (src == dest) ? doPayWithoutKyber(payData) : doPayWithKyber(payData);
+
+        // log as event
+        ProofOfPayment(destAddress, dest, paidAmount, paymentData);
```
Comment (yaronvel) [inline contracts/KyberPayWrapper.sol]: `beneficiary` is actually `msg.sender`.

File: contracts/KyberPayWrapper.sol [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        // verify the amount the user got is same as returned from Kyber Network
+        require(destAddressBalanceAfter > destAddressBalanceBefore);
+        require(paidAmount == (destAddressBalanceAfter - destAddressBalanceBefore));
+
+        // calculate the returned change amount
+        require(wrapperSrcBalanceBefore >= wrapperSrcBalanceAfter);
+        srcAmountUsed = wrapperSrcBalanceBefore - wrapperSrcBalanceAfter;
+
+        require(payData.srcAmount >= srcAmountUsed);
+        returnAmount = payData.srcAmount - srcAmountUsed;
```
Comment (yaronvel) [inline contracts/KyberPayWrapper.sol:141]: is this really true?

File: contracts/KyberPayWrapper.sol [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+        // return to sender the returned change
+        if (returnAmount > 0) {
+            if (payData.src == ETH_TOKEN_ADDRESS) msg.sender.transfer(returnAmount);
+            else {
```
Comment (yaronvel) [inline contracts/KyberPayWrapper.sol]: why not in one line?

File: contracts/KyberPayWrapper.sol [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        );
+    }
+
+    function getBalances (ERC20 src, ERC20 dest, address destAddress)
+        internal
+        view
+        returns (uint wrapperSrcBalance, uint destAddressBalance)
+    {
+        if (src == ETH_TOKEN_ADDRESS) wrapperSrcBalance = address(this).balance;
+        else {
```
Comment (yaronvel) [inline contracts/KyberPayWrapper.sol]: why {} only after `else`?

File: contracts/KyberPayWrapper.sol (hunk 1 above)
Comment (talbaneth) [reply contracts/KyberPayWrapper.sol:26]: as we saw it is not.

File: contracts/KyberPayWrapper.sol (hunk 2 above)
Comment (talbaneth) [reply contracts/KyberPayWrapper.sol]: done.

File: contracts/KyberPayWrapper.sol (hunk 3 above)
Comment (talbaneth) [reply contracts/KyberPayWrapper.sol]: as discussed will now have payer and payee in the event.

File: contracts/KyberPayWrapper.sol (hunk 4 above)
Comment (talbaneth) [reply contracts/KyberPayWrapper.sol:141]: as we talked it is.

File: contracts/KyberPayWrapper.sol (hunk 5 above)
Comment (talbaneth) [reply contracts/KyberPayWrapper.sol]: as we saw it causes linter problems.
changed it to use curly braces.

File: contracts/KyberPayWrapper.sol (hunk 6 above)
Comment (talbaneth) [reply contracts/KyberPayWrapper.sol]: same as above.

**Comment by talbaneth** [comment]
handled issues as we discussed.

**Pull request merged**